rollout/trafficrouting/plugin/rpc/rpc.go

```diff
@@ -0,0 +1,280 @@
+package rpc
+
+import (
+	"encoding/gob"
+	"fmt"
+	"net/rpc"
+
+	"github.com/argoproj/argo-rollouts/utils/plugin/types"
+
+	"github.com/argoproj/argo-rollouts/pkg/apis/rollouts/v1alpha1"
+	"github.com/hashicorp/go-plugin"
+)
+
+type UpdateHashArgs struct {
+	Rollout                v1alpha1.Rollout
+	CanaryHash             string
+	StableHash             string
+	AdditionalDestinations []v1alpha1.WeightDestination
+}
+
+type SetWeightAndVerifyWeightArgs struct {
+	Rollout                v1alpha1.Rollout
+	DesiredWeight          int32
+	AdditionalDestinations []v1alpha1.WeightDestination
+}
+
+type SetHeaderArgs struct {
+	Rollout        v1alpha1.Rollout
+	SetHeaderRoute v1alpha1.SetHeaderRoute
+}
+
+type SetMirrorArgs struct {
+	Rollout        v1alpha1.Rollout
+	SetMirrorRoute v1alpha1.SetMirrorRoute
+}
+
+type RemoveManagedRoutesArgs struct {
+	Rollout v1alpha1.Rollout
+}
+
+type VerifyWeightResponse struct {
+	Verified bool
+	Err      types.RpcError
+}
+
+func init() {
+	gob.RegisterName("UpdateHashArgs", new(UpdateHashArgs))
+	gob.RegisterName("SetWeightAndVerifyWeightArgs", new(SetWeightAndVerifyWeightArgs))
+	gob.RegisterName("SetHeaderArgs", new(SetHeaderArgs))
+	gob.RegisterName("SetMirrorArgs", new(SetMirrorArgs))
+	gob.RegisterName("RemoveManagedRoutesArgs", new(RemoveManagedRoutesArgs))
+}
+
+// TrafficRouterPlugin is the interface that we're exposing as a plugin. It needs to match metricproviders.Providers but we can
+// not import that package because it would create a circular dependency.
+type TrafficRouterPlugin interface {
+	InitPlugin() types.RpcError
+	types.RpcTrafficRoutingReconciler
+}
+
+// TrafficRouterPluginRPC Here is an implementation that talks over RPC
+type TrafficRouterPluginRPC struct{ client *rpc.Client }
+
+// NewTrafficRouterPlugin this is the client aka the controller side function that calls the server side rpc (plugin)
+// this gets called once during startup of the plugin and can be used to set up informers or k8s clients etc.
+func (g *TrafficRouterPluginRPC) InitPlugin() types.RpcError {
+	var resp types.RpcError
+	err := g.client.Call("Plugin.InitPlugin", new(interface{}), &resp)
+	if err != nil {
+		return types.RpcError{ErrorString: err.Error()}
```

Comment:
Does it make sense to add context to the RpcError? It would be interesting to know which call failed.